go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -38,77 +39,108 @@ import (
 
 	onDemandDuration            time.Duration
 	onDemandMu                  sync.Mutex
 	concurrentHeartbeatRequests int64
-	onDemandRequestTicks        int64
-	onDemandLastRequestTick     int64
 }
 
 // newHeartbeatWriter creates a new heartbeatWriter.
 func newHeartbeatWriter(env tabletenv.Env, alias *topodatapb.TabletAlias) *heartbeatWriter {
 	config := env.Config()
 
-	// config.EnableLagThrottler is a feature flag for the throttler; if throttler runs, then heartbeat must also run
-	if config.ReplicationTracker.Mode != tabletenv.Heartbeat && config.ReplicationTracker.HeartbeatOnDemand == 0 {
-		return &heartbeatWriter{}
+	configType := HeartbeatConfigTypeNone
+	onDemandDuration := defaultOnDemandDuration
+	switch {
+	case config.ReplicationTracker.HeartbeatOnDemand > 0:
+		configType = HeartbeatConfigTypeOnDemand
+		onDemandDuration = config.ReplicationTracker.HeartbeatOnDemand
+	case config.ReplicationTracker.Mode == tabletenv.Heartbeat:
+		configType = HeartbeatConfigTypeAlways
+		onDemandDuration = 0
 	}
 	heartbeatInterval := config.ReplicationTracker.HeartbeatInterval
+	if heartbeatInterval == 0 {
+		heartbeatInterval = defaultHeartbeatInterval
+	}
 	w := &heartbeatWriter{
 		env:              env,
-		enabled:          true,
+		configType:       configType,
 		tabletAlias:      alias.CloneVT(),
 		now:              time.Now,
 		interval:         heartbeatInterval,
-		onDemandDuration: config.ReplicationTracker.HeartbeatOnDemand,
+		onDemandDuration: onDemandDuration,
 		ticks:            timer.NewTimer(heartbeatInterval),
 		errorLog:         logutil.NewThrottledLogger("HeartbeatWriter", 60*time.Second),
 		// We make this pool size 2; to prevent pool exhausted
 		// stats from incrementing continually, and causing concern
 		appPool:      dbconnpool.NewConnectionPool("HeartbeatWriteAppPool", env.Exporter(), 2, mysqlctl.DbaIdleTimeout, 0, mysqlctl.PoolDynamicHostnameResolution),
 		allPrivsPool: dbconnpool.NewConnectionPool("HeartbeatWriteAllPrivsPool", env.Exporter(), 2, mysqlctl.DbaIdleTimeout, 0, mysqlctl.PoolDynamicHostnameResolution),
 	}
+
 	w.writeConnID.Store(-1)
-	if w.onDemandDuration > 0 {
-		// see RequestHeartbeats() for use of onDemandRequestTicks
-		// it's basically a mechanism to rate limit operation RequestHeartbeats().
-		// and selectively drop excessive requests.
-		w.allowNextHeartbeatRequest()
-		go func() {
-			// this will allow up to 1 request per (w.onDemandDuration / 4) to pass through
-			tick := time.NewTicker(w.onDemandDuration / 4)
-			defer tick.Stop()
-			for range tick.C {
-				w.allowNextHeartbeatRequest()
-			}
-		}()
-	}
 	return w
 }
 
+func (w *heartbeatWriter) unsafeOnDemandRequestsRateLimiter() *unsafe.Pointer {
+	return (*unsafe.Pointer)(unsafe.Pointer(&w.onDemandRequestsRateLimiter))
+}
+
+func (w *heartbeatWriter) safeOnDemandRequestsRateLimiter() *timer.RateLimiter {
+	result := (*timer.RateLimiter)(atomic.LoadPointer(w.unsafeOnDemandRequestsRateLimiter()))
+	return result
+}
```

Comment:
Can you please share why this was necessary? We should add function comments to lay that out as well. I'm curious about this. 🙂 It doesn't strike me as a necessary case for using the unsafe package. Would it not be better to use `onDemandRequestsRateLimiter atomic.Pointer[timer.RateLimiter]`?

Reply:
Yes! Good point, I will add comments.

Reply:
Switched to using `atomic.Pointer` and so happily removed these two functions.
FWIW the issue was that there could be a race condition:

- Someone issues a `RequestHeartbeats`, which uses the rate limiter
- The writer is being just `Close()`d